### PR TITLE
Round12: 1-epoch LR warmup before cosine decay (Lion stability)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Adding a 1-epoch linear learning rate warmup before the cosine annealing schedule will improve training stability and final performance. The SOTA baseline (PR #115) starts cosine annealing immediately from lr=1e-4. A warmup period allows gradients to stabilize before the optimizer receives the full learning rate signal, reducing early noise that can persist throughout training. This is especially relevant with Lion optimizer, which has more aggressive gradient updates than Adam.

Reference: "Cyclical Learning Rates for Training Neural Networks" (Smith, 2017); Lion warmup analysis in "Symbolic Discovery of Optimization Algorithms" (Chen et al., 2023) shows Lion benefits from a warmup period.

## Instructions

Start from the SOTA baseline stack and add `--lr-warmup-epochs 1`:

```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --optimizer lion \
  --lr 1e-4 \
  --weight-decay 5e-4 \
  --no-compile-model \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --model-slices 128 \
  --ema-decay 0.999 \
  --lr-warmup-epochs 1 \
  --wandb_group tay-round12-lr-warmup-1ep
```

**One change from baseline**: `--lr-warmup-epochs 1` added. All other hyperparameters are identical to SOTA.

After training completes, report:
- `val_primary/abupt_axis_mean_rel_l2_pct` (primary metric, lower is better)
- `full_val_primary/abupt_axis_mean_rel_l2_pct` (full validation)
- Per-axis breakdown: surface_pressure, volume_pressure, wall_shear_x/y/z
- W&B run ID and link

## Baseline

Current SOTA — PR #115 (thorfinn), W&B run `d03oghpp`:
- **val_abupt**: 9.484%
- **test_abupt**: 10.580%
- surface_pressure: 5.690%
- volume_pressure: 12.740%
- wall_shear_x: 8.908%
- wall_shear_y: 12.491%
- wall_shear_z: 13.071%

Reproduce baseline:
```bash
cd target/
torchrun --standalone --nproc_per_node=8 train.py \
  --agent thorfinn --optimizer lion --lr 1e-4 --weight-decay 5e-4 \
  --no-compile-model --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --ema-decay 0.999
```

**Target to beat**: val_abupt < 9.484%

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Early W&B reading at step 6038: val_abupt=41.93% — notably ahead of SOTA ep1 pace (53.75%), suggesting the 1-epoch warmup is helping early convergence. Steep descent slope of -9.48pp/1k is healthy.

One flag to verify: the PR body uses `--wandb_group` (underscore) rather than `--wandb-group` (hyphen). The run *is* appearing under group `tay-round12-lr-warmup-1ep` correctly in W&B (run `ut1qmc3i`), so this is likely fine — but please confirm that is the case and that the run is grouping correctly.

Keep running to ep9 and report final metrics. Looking forward to seeing if this advantage persists through training.

**Comment by morganmcg1** [comment]
STUDENT fern: Confirmed wandb grouping is correct. The actual command running uses `--wandb-group tay-round12-lr-warmup-1ep` (hyphen, matching argparse). The PR body's `--wandb_group` (underscore) was a typo I corrected at launch. W&B run `ut1qmc3i` is correctly grouped under `tay-round12-lr-warmup-1ep`.

Training is healthy — currently at Epoch 3/50, ~62% through. Will report final metrics when training completes.

**Comment by morganmcg1** [comment]
ADVISOR: Strong trajectory — currently the best active experiment.

Curve so far:
- ep1=67.726% (high — expected with 1ep warmup; LR is still ramping)
- ep2=41.929%
- ep3=19.303%
- ep4=13.733%
- ep5=11.602%
- ep6=10.502%
- ep7=9.876% (delta=+0.392% vs SOTA val 9.484%)

ep7 is within 0.4pp of SOTA val and the slope is still steep. ep8/ep9 could plausibly cross — and a 1ep warmup over 9-epoch training is genuinely under-explored vs the flat-LR baseline.

**Do not stop early.** Run all 9 epochs and capture the test eval from the best val checkpoint (not terminal epoch). If best val < 9.484%, this is a probable merge candidate even with a small margin.

Watching closely.

**Comment by morganmcg1** [comment]
ADVISOR: **ep9 confirmed — 9.2910% beats SOTA, run complete.**

Full trajectory for W&B run `ut1qmc3i` (fern rank0, group `tay-round12-lr-warmup-1ep`):

| Epoch | Step  | val_abupt | surf_pres | vol_pres | wall_shear |
|-------|-------|-----------|-----------|----------|------------|
| 1     | 2720  | 67.7263%  | 52.1452%  | 59.6851% | 68.9568%   |
| 2     | 5441  | 41.9288%  | 31.5194%  | 25.0452% | 46.0780%   |
| 3     | 8162  | 19.3033%  | 13.3635%  | 11.4773% | 21.4948%   |
| 4     | 10883 | 13.7327%  | 9.2451%   | 8.0774%  | 15.3581%   |
| 5     | 13604 | 11.6016%  | 7.6196%   | 6.8622%  | 13.0241%   |
| 6     | 16325 | 10.5020%  | 6.7791%   | 6.2569%  | 11.7950%   |
| 7     | 19046 | 9.8759%   | 6.3077%   | 6.0145%  | 11.0603%   |
| 8     | 21767 | 9.4516%   | 6.0019%   | 5.7614%  | 10.5847%   |
| **9** | **23544** | **9.2910%** | **5.8707%** | **5.8789%** | **10.3423%** |

**Best ep9 val_abupt = 9.2910%** — beats SOTA 9.484% by **-0.193pp (-2.03%)**.

The lr_warmup_epochs=1 hypothesis is confirmed positive. Smooth early descent, continuous improvement across all 9 epochs, no instability. Proceeding to merge.
